src/orange/detail/BIHIntersectingVolFinder.hh

```diff
@@ -0,0 +1,260 @@
+
+//---------------------------------------------------------------------------//
+/*!
+ * Calculate the minimum intersection.
+ */
+template<class F>
+CELER_FUNCTION auto
+BIHIntersectingVolFinder::operator()(BIHIntersectingVolFinder::Ray const& ray,
+                                     F&& visit_vol) const -> Intersection
+{
+    BIHNodeId previous_node;
+    BIHNodeId current_node{0};
+
+    Intersection intersection{OnLocalSurface{},
+                              std::numeric_limits<real_type>::infinity()};
+
+    do
+    {
+        if (!view_.is_inner(current_node))
+        {
+            intersection = this->visit_leaf(
+                view_.leaf_node(current_node), ray, intersection, visit_vol);
+        }
+
+        previous_node = exchange(
+            current_node,
+            this->next_node(
+                current_node, previous_node, ray, intersection.distance));
+
+    } while (current_node);
+
+    return this->visit_inf_vols(intersection, visit_vol);
+}
+
+//---------------------------------------------------------------------------//
+// HELPER FUNCTIONS
+//---------------------------------------------------------------------------//
+/*!
+ *  Get the ID of the next node in the traversal sequence.
+ */
+CELER_FUNCTION
+BIHNodeId BIHIntersectingVolFinder::next_node(BIHNodeId const& current_id,
+                                              BIHNodeId const& previous_id,
+                                              Ray const& ray,
+                                              double min_dist) const
+{
+    using Side = BIHInnerNode::Side;
+
+    BIHNodeId next_id;
+
+    if (view_.is_inner(current_id))
+    {
+        auto const& current_node = view_.inner_node(current_id);
+        auto const& l_edge = current_node.edges[Side::left];
+        auto const& r_edge = current_node.edges[Side::right];
+
+        if (previous_id == current_node.parent)
+        {
+            // Visiting this inner node for the first time; go down either left
+            // or right edge
+            next_id = this->visit_bbox(l_edge.bbox, ray, min_dist)
+                          ? l_edge.child
+                          : r_edge.child;
+        }
+        else if (previous_id == current_node.edges[Side::left].child)
+        {
+            // Visiting this inner node for the second time; go down right edge
+            // or return to parent
+            next_id = this->visit_bbox(r_edge.bbox, ray, min_dist)
+                          ? r_edge.child
+                          : current_node.parent;
+        }
+        else
+        {
+            // Visiting this inner node for the third time; return to parent
+            CELER_EXPECT(previous_id == current_node.edges[Side::right].child);
+            next_id = current_node.parent;
+        }
+    }
+    else
+    {
+        // Leaf node; return to parent
+        CELER_EXPECT(previous_id == view_.leaf_node(current_id).parent);
+        next_id = previous_id;
+    }
+
+    return next_id;
```

Comment:
I've waffled over the years about the clarity of having a single "return" versus a longer scope and if/else, but I think this function would read better without the local `next_id` and just:

```c++
    if (!view_.is_inner(current_id))
    {
        CELER_EXPECT(previous_id == view_.leaf_node(current_id).parent);
        return previous_id;
    }

    auto const& current_node = view_.inner_node(current_id);
    auto const& l_edge = current_node.edges[Side::left];
    auto const& r_edge = current_node.edges[Side::right];

    if (previous_id == current_node.parent)
    {
        // Visiting this inner node for the first time; go down either left
        // or right edge
        return this->visit_bbox(l_edge.bbox, ray, min_dist) ? l_edge.child
                                                            : r_edge.child;
    }

    if (previous_id == current_node.edges[Side::left].child)
    {
        // Visiting this inner node for the second time; go down right edge
        // or return to parent
        return this->visit_bbox(r_edge.bbox, ray, min_dist)
                   ? r_edge.child
                   : current_node.parent;
    }

    // Visiting this inner node for the third time; return to parent
    CELER_ASSERT(previous_id == current_node.edges[Side::right].child);
    next_id = current_node.parent;
```